✨ Feature : 피드 공유하기 기능

kmss6905 opened the pull request:
## 💻 작업 내역 
### ✓ 피드 공유하기 기능 1b01714e6204eae72701bfadeee57e2e4b49a635

피드 공유하기 로직은 아래와 같습니다.

1. 피드 번호로 피드 조회
2. 피드에 있는 SNS 에 종류에 따라 공유 요청을 처리할 수 있는 적절한 SnsClient 찾기
3. 외부 SNS에 공유하기 요청
4. 피드 공유 내역 저장
5. 피드 공유 링크를 포함한 응답값 반환


### ✓ 외부 클라이언트 응답값 변경 f5296fbe59f4ee1a8ae885c4adbde1d421ca2a1c 

외부 API에 공유하기를 요청할 때 공유 링크를 반환한다고 가정하였습니다.


## 🔎 PR 특이 사항

develop 브랜치에 ErrorType 클래스에 동일한 에러코드가 있어서 깨졌던 문제 해결했습니다. 6771f56cef620a56546489a8f95c08f47ead2da9


피드 엔티티 수정하였습니다 - 충돌유의 5a53ec65cb3a59c90f05b456558033dd7cb64d5d
- 카운트 필드 및 카운트 증가 메서드 추가
@min050410 @hsjkdss228

---
**추가된 내용**

✅  응답으로 공유 링크 제공

보통 SNS 게시글을 공유 했을 때 복사할 수 있는 링크를 제공합니다.
그렇기 때문에 외부 API 를 통해 공유링크(url)를 가져오도록 하였습니다. 
이를 통해 외부에 등록되어 있는 각 SNS 피드로 바로 이동할 수 있도록 하였습니다. 

응답

```json

{
	"code": 200,
	"data": {
		"url": "https://www.instagram.com/feeds/:contentId"
	}
}

```

따라서 공유 카운트를 증가시키되(내부에서는 저장), 해당 피드가 있는 곳으로 바로 접근할 수 있도록 합니다.
실제 해당 피드가 위치한 곳으로 갈 수 있도록 하는 링크의 경우 가상으로 구성하였습니다
(ex, `https://www.instagram.com/feeds/{contentId}` )

kmss6905 committed:
Bugfix: ErrorType 병합 잘못된 부분 수정

kmss6905 committed:
Feature: 공유(Share) 엔티티 추가

- 잘못된 공유 객체 생성 시 던지는 예외 클래스 추가(`ShareFeedIdNotNullException.class`, `ShareUserIdNotNullException.class`)
- ErrorType 에러 코드 추가

kmss6905 committed:
Test: 공유(Share) 엔티티 테스트

kmss6905 committed:
Feature: SNS 클라이언트 공유 기능 수정

- 공유 링크도 같이 응답하도록 변경

kmss6905 committed:
Test: SNS 클라이언트 공유 요청 테스트

kmss6905 committed:
Feature: 피드 공유 서비스 추가

- FeedShareResponseDto 추가
- ShareRepository 추가
- ErrorType 에러코드 및 공유 실패 예외 클래스 추가
- swagger tag 피드 추가

kmss6905 committed:
Test: 피드 공유 서비스 테스트

kmss6905 committed:
Merge branch 'develop' of github.com:wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service into Feature/FEED-20

- ErrorType 충돌해결

buhee1029 commented on src/main/java/com/wanted/feed/feign/SnsClient.java:
```diff
@@ -14,12 +15,9 @@ default ResponseEntity<String> likeFeed(@PathVariable String contentId) {
-        verificationNullContentId(contentId);
-        return ResponseEntity.ok("ok");
-    }
+    ResponseEntity<ClientShareResponseDto> shareFeed(@PathVariable String contentId);
```
공통 응답 포맷 ApiResponse<T> 를 사용하지 않은 이유가 있을까요?

kmss6905 replied:
이 부분은 고민을 많이 했던 것 같아요 🤔
저는 실제 클라이언트로 가는 공통 응답 포맷을 외부 API 를 받을 때 사용하는 포맷을 동일하게 하면 클라이언트로 가는 API 와 착각할 수 있을 것 같아서 기본으로 제공하는 응답 포맷을 사용하였습니다.

buhee1029 reviewed (commented):
수고하셨습니다 ~😆 좋아요 기능 때부터 느꼈는데 꼼꼼한 테스트 좋은 것 같아요 :)

kmss6905 committed:
Feature: SNS 클라이언트 공유 응답값 수정

-`https://www.instagram.com/feeds/{contendId}` -> `https://www.instagram.com/feeds/{contendId}` 로 변경

kmss6905 committed:
Feature: 피드 엔티티 수정

- `shareCount`, `likeCount` 필드 추가

kmss6905 committed:
Feature: 피드 공유 서비스 수정

- 공유 저장 시 피드 내 공유 카운트 1 증가

kmss6905 committed:
Test: 피드 공유 서비스 테스트 수정

kmss6905 committed:
Merge branch 'develop' of github.com:wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service into feature/FEED-20

- ErrorType 충돌 해결
- 사용하지 않은 import 제거
- 기존 스켈레톤에서 가져온 의미없는 에러코드 제거
